Partition Simplification Framework

gselzer opened the pull request:
Solving issues like scijava/scijava#146 and acting on scijava/scijava#36 requires us to gain some insight about the forms of types and to coalesce structurally equivalent Ops, and there's a lot of overlap with the simplification framework. Unfortunately, the simplified Ops and types are hidden within the simplification framework.

This PR partitions `SimplifiedOpInfo` into two separate `OpInfo` implementations:
* `SimplifiedOpInfo`s consist of an `OpInfo` along with `focus` Ops to focus the inputs and a `simplify` Op to simplify the output. For example, if the original `OpInfo` is `Double output = foo.bar(Double d1, Double d2)`, the simplified `OpInfo` might be `Number output = foo.bar(Number d1, Number d2)`.
* `FocusedOpInfo` takes a `SimplifiedOpInfo` along with `simplify` Ops to simplify the inputs *to the simple type* and `focus` Ops to focus the output *from the simple type*. To continue the above example, a focused `OpInfo` might be `Integer output = foo.bar(Integer d1, Integer d2)`.

Note that this PR is built atop of #94, to make use of the new `Ops` utility class.

gselzer committed:
Partition simplification

This allows us to access simple descriptions, which we now do in this
commit

ctrueden commented on imagej/imagej-ops2/src/test/java/net/imagej/ops2/create/CreateImgTest.java:
```diff
@@ -32,9 +32,11 @@
 import java.util.function.BiFunction;
 import java.util.function.Function;
 
+import com.google.errorprone.annotations.FormatString;
```
Looks like an unintended import?

gselzer replied:
Oop, thanks! Removed now

ctrueden commented on scijava/scijava-ops-engine/src/main/java/module-info.java:
```diff
@@ -61,6 +92,9 @@
 	    org.scijava.ops.engine.impl.OpCollectionInfoGenerator,
 		org.scijava.ops.engine.matcher.reduce.ReducedOpInfoGenerator;
 
+	provides org.scijava.ops.engine.OpDescriptionGenerator with
```
Maybe this interface should be part of the SPI module instead?

gselzer replied:
I can't say I'm a huge fan of that - the goal is for a library like Imglib2-algorithm to extend the SPI, not the API (although of course now we probably just want it to use javadoc, but that's the kind library). I don't see that kind of library also implementing an `OpDescriptionGenerator` - i.e. I think it aligns more with the `MatchingRoutine` interface in the API than the `OpCollection` interface in the SPI.

ctrueden commented on scijava/scijava-ops-api/src/main/java/org/scijava/ops/api/OpBuilder.java:
```diff
@@ -91,67 +92,67 @@ public OpBuilder(final OpEnvironment env, final String opName, final Hints hints
+         * Finds all Ops matching the current partial OpBuilder request
+         * @return a verbose {@link String} for each Op satisfying the partial request
+         */
+		public List<String> helpVerbose () {
```
`helpVerbose ()` &rarr; `helpVerbose()` here and everywhere

gselzer replied:
Messed up the template - fixed now!

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/OpDescription.java:
```diff
@@ -105,4 +121,28 @@ private static void appendParam(final StringBuilder sb, final Member<?> arg,
         sb.append("\n");
     }
+
+    public static String typeString(final Type input, final boolean simple) {
```
Can this be private? AFAICT it is only called here in the same class.

gselzer replied:
Fixed!

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/OpDescriptionGenerator.java:
```diff
@@ -0,0 +1,24 @@
+
```
License header

gselzer replied:
Added!

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/impl/DefaultOpEnvironment.java:
```diff
@@ -98,9 +101,9 @@
  */
```
Probably overdue to add `@author Gabriel Selzer` to this class :laughing:

gselzer replied:
Added

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/impl/DefaultOpEnvironment.java:
```diff
@@ -146,24 +149,17 @@ public class DefaultOpEnvironment implements OpEnvironment {
 		this.typeService = new DefaultTypeReifier(Discoverer.using(ServiceLoader::load));
 		this.history = OpHistory.getOpHistory();
-		matcher = new DefaultOpMatcher(Discoverer.using(ServiceLoader::load).discover(
+		this.matcher = new DefaultOpMatcher(Discoverer.using(ServiceLoader::load).discover(
```
Ooh, that's a style Pandora's box you're opening! Are you sure you want to do that? Refer to instance fields with `this.` all the time? It would change a lot of code!

gselzer replied:
Fixed

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/impl/DefaultOpEnvironment.java:
```diff
@@ -315,32 +315,67 @@ public <T> T bakeLambdaType(T op, Type type) {
+
+	private List<OpInfo> generateAllInfos(Object o) {
+		return ServiceLoader.load(OpInfoGenerator.class).stream() //
+				.map(ServiceLoader.Provider::get) //
```
Does this relate back to that JPMS visibility hack we discovered before? I think a comment is in order here to explain what the `Object o` here is, what the `ServiceLoader.Provider::get` is doing, etc.

gselzer replied:
Fixed

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/impl/DefaultOpEnvironment.java:
```diff
@@ -315,32 +315,67 @@ public <T> T bakeLambdaType(T op, Type type) {
+	 * {@link OpInfo}. This is consistent with {@link DefaultOpEnvironment#infos}
+	 * returning multiple instances of the same {@link OpInfo}. The duplicate
+	 * {@link OpInfo}s are created when Ops have multiple names.
+	 *
```
Should properly have a `</p>` at the end of this paragraph.

gselzer replied:
Added

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/impl/DefaultOpEnvironment.java:
```diff
@@ -649,7 +689,6 @@ private RichOp<?> wrap(OpInstance<?> instance, Hints hints) {
 	 * @param type the functional {@link Type} of the {@code op} we're looking for
 	 * @param name the name of the {@code op} we're looking for
 	 * @param typeVarAssigns the mappings of {@link TypeVariable}s to {@link Type}s 
-	 * @return null if the specified type has no functional method
```
Maybe there should be a `@throws OpMatchingException` explanation of that case? I know, it's private, but still. The javadoc was taken this far! May as well take it all the way.

gselzer replied:
Added it!

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/impl/DefaultOpEnvironment.java:
```diff
@@ -682,67 +721,29 @@ private OpRequest inferOpRequest(Type type, String name, Map<TypeVariable<?>, Ty
-	private synchronized void initOpDirectory() {
-		if (opDirectory != null) return;
-		Map<String, List<OpInfo>> tmp = new HashMap<>();
-		// add all OpInfos that are directly discoverable
```
Wow, this was a dense thicket of behavior. If it were staying, I would demand comments here.

I'm not exactly seeing where this logic went in the new code, though...

gselzer replied:
I got rid of it! The `initOpDirectory` method allowed us to lazily initialize the `Map<String, SortedSet<OpInfo>> opDirectory`, but now we eagerly initialize it, so no need for this functionality.

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/impl/DefaultOpEnvironment.java:
```diff
@@ -682,67 +721,29 @@ private OpRequest inferOpRequest(Type type, String name, Map<TypeVariable<?>, Ty
 		}
 		for (String opName : opInfo.names()) {
-			directory.computeIfAbsent(opName, name -> new ArrayList<>()).add(opInfo);
+			opDirectory.computeIfAbsent(opName, name -> new TreeSet<>()).add(opInfo);
```
Is the use of `TreeSet` so that it always stays sorted?

gselzer replied:
> Is the use of `TreeSet` so that it always stays sorted?

@ctrueden yes, I want it sorted by Op priority - I'm starting to think that maybe instead of `List<OpInfo>` being returned by `OpEnvironment.infos` we should return a `SortedSet<OpInfo>` - that way, you'd get two guarantees:
1. The sorting gives you Ops by priority
2. The set prevents duplicates.

ctrueden replied:
Yes, `SortedSet` seems like the right interface here! :+1:

gselzer replied:
Changed things to use `SortedSet` across the board.

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/FocusedInfoTreeGenerator.java:
```diff
@@ -0,0 +1,136 @@
+import org.scijava.types.Nil;
+import org.scijava.types.Types;
+
+public class FocusedInfoTreeGenerator implements InfoTreeGenerator {
```
javadoc

gselzer replied:
Added!

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/FocusedInfoTreeGenerator.java:
```diff
@@ -0,0 +1,136 @@
+		return new InfoTree(baseInfo, originalChain.dependencies());
+	}
+
+	List<String> parseComponents(String signature) {
```
Should be `private` (as should the version of this function in `SimplifiedInfoTreeGenerator`)

gselzer replied:
Changed it!

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/FocusedOpInfo.java:
```diff
@@ -0,0 +1,366 @@
+
```
license header

gselzer replied:
Added!

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/FocusedOpInfo.java:
```diff
@@ -0,0 +1,366 @@
+import org.scijava.types.Nil;
+import org.scijava.types.Types;
+
+public class FocusedOpInfo implements OpInfo {
```
javadoc

gselzer replied:
Added!

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/FocusedOpInfo.java:
```diff
@@ -0,0 +1,366 @@
+	 * infinite loop (and eventually a {@link StackOverflowError}. This means that
+	 * we cannot find a simplified copy Op <b>unless a direct match can be
+	 * found</b>, at which point we might as well just use the direct match.
+	 * <p>
```
These paragraphs should be terminated with `</p>`.

gselzer replied:
Fixed!

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/FocusedOpInfo.java:
```diff
@@ -0,0 +1,366 @@
+				.get(i)));
+		}
+
+		// TODO: only calculate the loss once
```
I don't understand this TODO. The following call processes the output/outputType, whereas the above loop processes each input. Does it mean to combine the loop + output calculations into a single function call somehow? Would this make the framework significantly faster? If not, I'd say kill the TODO comment.

gselzer replied:
Uh, I think the point was that we want to avoid computing the priority every time we call `FocusedOpInfo.getPriority()`, but I have no idea why the `TODO` was there.

Since there isn't a major harm in computing it more than once, I just added a `final Double priority = null` field to `FocusedOpInfo`, and whenever `FocusedOpInfo.getPriority()` is called we compute the priority if `priority == null`. I think that's enough to kill the `TODO`

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/Identity.java:
```diff
@@ -35,9 +35,10 @@
+@OpClass(names = "simplify, focus, identity", priority = Priority.LAST)
 public class Identity<T> implements Function<T, T>, Op {
 
 	public Identity() {}
```
This constructor is redundant, no?

gselzer replied:
Yup, removed it!

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/Identity.java:
```diff
@@ -35,9 +35,10 @@
 
 @OpHints(hints = { Simplification.FORBIDDEN })
-@OpClass(names = "simplify, focus")
+@OpClass(names = "simplify, focus, identity", priority = Priority.LAST)
```
Why does this op need to be an `OpClass` rather than a more concise `OpField` using a lambda?

gselzer replied:
I mean, you'll still have the `OpCollection` boilerplate, but just for you I made it into a `OpField`!

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/IdentityLossReporter.java:
```diff
@@ -43,16 +44,16 @@
-@OpClass(names = "lossReporter")
-public class IdentityLossReporter<T> implements LossReporter<T, T>, Op {
+@OpClass(names = "lossReporter", priority= Priority.VERY_HIGH)
+public class IdentityLossReporter<U, T extends U> implements LossReporter<T, U>, Op {
```
Why does this op need to be an `OpClass` rather than a more concise `OpField` using a lambda?

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/PrimitiveArraySimplifiers.java:
```diff
@@ -44,63 +45,106 @@
+	@OpHints(hints = { Simplification.FORBIDDEN })
 	@OpField(names = "focus")
 	public final Function<ObjectArray<Number>, Byte[]> byteArrFocuser = o -> o
 		.stream().map(b -> b == null ? null : b.byteValue()).toArray(Byte[]::new);
```
Why are we using `Byte[]` rather than `byte[]` here?

gselzer replied:
Uh, no reason, we should have both though, which I have thus added!

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/PrimitiveArraySimplifiers.java:
```diff
@@ -44,63 +45,106 @@
-	@OpHints(hints = {Simplification.FORBIDDEN})
+	@OpHints(hints = { Simplification.FORBIDDEN })
 	@OpField(names = "simplify")
 	public final Function<Short[], ObjectArray<Number>> shortArrSimplifier =
```
Same question: why `Short[]` and not `short[]`?

gselzer replied:
Uh, no reason, we should have both though, which I have thus added!

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/PrimitiveArraySimplifiers.java:
```diff
@@ -44,63 +45,106 @@
+
+	@OpHints(hints = { Simplification.FORBIDDEN })
 	@OpField(names = "focus")
 	public final Function<ObjectArray<Number>, Double[]> doubleArrFocuser = o -> o
```
Why `Double[]` and not `double[]`?

gselzer replied:
Uh, no reason, we should have both though, which I have thus added!

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/SimplifiedInfoTreeGenerator.java:
```diff
@@ -32,54 +32,40 @@
+import org.scijava.types.Nil;
 
-public class SimplificationInfoTreeGenerator implements InfoTreeGenerator {
+public class SimplifiedInfoTreeGenerator implements InfoTreeGenerator {
```
I like this rename.

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/SimplifiedOpDescriptionGenerator.java:
```diff
@@ -0,0 +1,53 @@
+
```
license header

gselzer replied:
Added

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/SimplifiedOpDescriptionGenerator.java:
```diff
@@ -0,0 +1,53 @@
+import org.scijava.ops.engine.OpDescriptionGenerator;
+import org.scijava.priority.Priority;
+
+public class SimplifiedOpDescriptionGenerator implements
```
Javadoc

gselzer replied:
Added!

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/SimplifiedOpInfo.java:
```diff
@@ -134,89 +135,14 @@ public double priority() {
 	@Override
 	public String implementationName() {
-		return srcInfo.implementationName() + " simplified to a " + opType();
+		return info.implementationName() + "_simple";
```
Instead of `_`, which is a legal character in Java class names, why not an invalid character like `|`, to ensure that there cannot be a name clash between `Foo_simple` the class and `Foo_simple` the simplified `Foo`?

gselzer replied:
I figured that this would conflict with `SimplifiedOpInfo.signature()`, but I guess it doesn't (which is great!), so I merged it!

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/SimplifiedOpInfo.java:
```diff
@@ -294,29 +226,51 @@ public String version() {
+	private Type[] inTypes(Type[] originalInputs, List<RichOp<Function<?, ?>>> inputFocusers) {
+		Map<TypeVariable<?>, Type> typeAssigns = new HashMap<>();
+		Type[] inTypes = new Type[originalInputs.length];
+		// TODO: This feels kind of inefficient, but we have to do each individually
```
Should this be an `NB:` prefix instead of `TODO:`? It is just explaining why the code is written this way...

gselzer replied:
Sure, changed!

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/SimplifiedOpRequest.java:
```diff
@@ -165,20 +172,17 @@ public String getName() {
 	public Type getType() {
-		throw new UnsupportedOperationException(
-			"The type of a SimplifiedOpRequest is indeterminate; it must be matched with a OpInfo to form a concrete Type");
+		return type;
```
:tada:

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/SimplifiedOpRequest.java:
```diff
@@ -191,12 +195,32 @@ public boolean typesMatch(Type opType) {
+			}
+		}
+		else {
+			if (!Types.isAssignable(opType, type)) {
```
Could collapse this `if` onto the enclosing `else` to reduce nesting.

gselzer replied:
Thanks, good idea!

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/util/Ops.java:
```diff
@@ -82,7 +82,7 @@ private Ops() {
 	 */
 	public static String[] parseOpNames(String names) {
-		return Arrays.stream(names.split(",")).map(s -> s.trim())
+		return Arrays.stream(names.split("\\s*,\\s*")).map(s -> s.trim())
```
Doesn't the `s.trim()` already take care of squelching any whitespace surrounding the commas?

gselzer replied:
...it should. I don't remember why I added it in...

Removed it, though.

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/util/Ops.java:
```diff
@@ -162,30 +162,41 @@ public static Class<?> findFunctionalInterface(Class<?> type) {
 	public static Method findFunctionalMethod(Class<?> cls) {
+		Method m = fMethod(cls);
+		if (m == null) {
+			throw new IllegalArgumentException("Op type" + cls +
```
`"Op type" + cls +` &rarr; `"Op type " + cls.getName() +`

(NB: The above suggestion fixes *two* bugs...)

gselzer replied:
Fixed!

ctrueden commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/util/Ops.java:
```diff
@@ -162,30 +162,41 @@ public static Class<?> findFunctionalInterface(Class<?> type) {
-		
+
+		List<Method> nonDefaults = Arrays.stream(iFace.getMethods()).filter(m -> !m
+			.isDefault()).collect(Collectors.toList());
```
This change is the victim of autoformatting, which made the line wrapping worse.

gselzer replied:
Fixed!

ctrueden commented on scijava/scijava-ops-engine/src/test/java/org/scijava/ops/engine/impl/ProvenanceTest.java:
```diff
@@ -358,6 +359,35 @@ public void testAdaptedOpWithDependencies() {
 	public void testSimplificationRecovery() {
+		// Get the Op
+		Computers.Arity1<ObjectArray<Number>, ObjectArray<Number>> c = ops //
+				.op("test" + ".provenanceComputer") //
```
Why is this string split with the `+`  operator here?

gselzer replied:
It isn't anymore!

ctrueden commented on scijava/scijava-ops-engine/src/test/java/org/scijava/ops/engine/impl/ProvenanceTest.java:
```diff
@@ -358,6 +359,35 @@ public void testAdaptedOpWithDependencies() {
+	public void testFocusedRecovery() {
 		// Get the Op
 		Computers.Arity1<Integer[], Integer[]> c = ops //
 			.op("test" + ".provenanceComputer") //
```
Why is this string split with the `+` operator here?

gselzer replied:
Fixed!

ctrueden commented on scijava/scijava-ops-engine/src/test/java/org/scijava/ops/engine/matcher/adapt/OpAdaptationInfoTest.java:
```diff
@@ -46,7 +46,7 @@ public class OpAdaptationInfoTest extends AbstractTestEnvironment {
 	@BeforeAll
 	public static void addNeededOps() {
-		ops.registerInfosFrom(new CopyOpCollection<>());
+		ops.register(new CopyOpCollection<>());
```
Every time I see this renamed method, I smile. :smile_cat:

ctrueden commented on scijava/scijava-ops-engine/src/test/java/org/scijava/ops/engine/matcher/simplify/SimplifyTest.java:
```diff
@@ -52,15 +52,17 @@ public class SimplifyTest extends AbstractTestEnvironment implements OpCollectio
+			new PrimitiveSimplifiers(), //
+			new PrimitiveLossReporters(), //
+			new IdentityLossReporter(), //
+			new Identity(), //
```
`Identity` should have a `<>`, no?

gselzer replied:
Added it!

ctrueden commented on scijava/scijava-types/src/main/java/org/scijava/types/Types.java:
```diff
@@ -1676,6 +1678,10 @@ private static boolean isAssignable(final Type type,
 				return true;
 			}
 
+			if (type.equals(Any.class)) {
```
`Any` is not `final`.... so this check would fail if someone extends it.... :wink: 

(To clarify: I think `Any` should probably be final—but if not, this line should change to something like `type instanceof Class && Any.class.isAssignableFrom((Class<?>) type)`.)

gselzer replied:
Made `Any` `final`

ctrueden commented on scijava/scijava-types/src/main/java/org/scijava/types/Types.java:
```diff
@@ -2239,7 +2246,7 @@ private static boolean isAssignable(final Type type,
 			}
 			
-			if (type instanceof Any) {
+			if (type instanceof Any || type.equals(Any.class)) {
```
Again, `Any` is not `final`...

gselzer replied:
It is now!

ctrueden reviewed (commented):
Oof, got through it! I skimmed some parts that were above my level of understanding, but hopefully it helps anyway.

gselzer committed:
Refactor Engine Ops class -> Infos

gselzer commented on imagej/imagej-ops2-tutorial/docs/SearchingForOps.md:
```diff
@@ -2,13 +2,14 @@
 As the `OpEnvironment` is fully extensible, different `OpEnvironment`s might contain different Ops, and it is important to be able to query an `OpEnvironment` about the available Ops.
 
-Users can query the `OpEnvironment` for Ops matching a given name using the method `OpEnvironment.descriptions(String)`. The following SciJava script queries an `OpEnvironment` for all Ops matching the name `filter.gauss`:
+Users can query the `OpEnvironment` for available Ops using the method `OpEnvironment.help(String)`. The following example shows you how to find all `"filter.gauss"` Ops in an `OpEnvironment`:
```
We should describe the new namespace searching functionality

gselzer replied:
First cut in 77539361

gselzer committed:
Remove List import in OpBuilder

gselzer commented on scijava/scijava-ops-api/src/main/java/org/scijava/ops/api/OpEnvironment.java:
```diff
@@ -818,29 +831,63 @@ default void registerInfosFrom(Object o) {
+	 * {@link OpEnvironment}
+	 *
+	 * @param request the {@link OpRequest} to filter on
+	 * @return a {@link Set} of descriptions
```
These `@return` statements are wrong

gselzer replied:
Cleaned up in c407ba2c

gselzer commented on scijava/scijava-ops-api/src/main/java/org/scijava/ops/api/OpRequest.java:
```diff
@@ -132,4 +135,68 @@ static void append(final StringBuilder sb, final String s) {
+		this.outType = Optional.of(outType.getType());
+	}
+
+	@Override public String getName() {
```
Could use a format

gselzer replied:
Formatted in 47fd17da

gselzer commented on scijava/scijava-ops-api/src/main/java/org/scijava/ops/api/OpRequest.java:
```diff
@@ -132,4 +135,68 @@ static void append(final StringBuilder sb, final String s) {
+
+class PartialOpRequest implements OpRequest {
+
+	private final Optional<String> name;
+	private final Optional<Type[]> args;
+	private final Optional<Type> outType;
```
If we're just going to return `null` when these are empty, might as well just set them to `null`...

gselzer replied:
Cleaned up (all of `OpRequest`) in 1b51740e

gselzer reviewed (commented):
Started out with a mini-review

gselzer commented on scijava/scijava-ops-api/src/main/java/org/scijava/ops/api/OpInfo.java:
```diff
@@ -114,6 +114,10 @@ default Type outputType() {
 	default int compareTo(final OpInfo that) {
 		if (this.priority() < that.priority()) return 1;
 		if (this.priority() > that.priority()) return -1;
+		int nameComparison = this.toString().compareTo(that.toString());
```
It might be nice to think about this a little more, and also to make sure that it aligns with the `equals` method...

hinerm replied:
@gselzer is simply checking `implementationName` insufficient for some reason? Is it actually possible for the `toString` result of two `OpInfo`s to be the same but the `implementationName` different?

gselzer replied:
@hinerm so the one case where we might want this is [here](https://github.com/scijava/incubator/pull/95/files#diff-af5dec84a4906407c0159a8ff35d0858d018c6a57302bb6e9c8d81d901107e47R77) - that test will fail because [`OpEnvironment.infos()`](https://github.com/scijava/incubator/pull/95/files#diff-b271cbe48149d64e6e0bfe70e70eadb7d5d3e3f0d83988a06c2e837af4831103R174-R176) will consider the two `OpInfo`s to be the same.

What if we instead changed the `List<String> names()` method of `OpInfo` to return a `Set<String>`, and then we could check set equality in `compareTo` and `equals`?

gselzer replied:
@hinerm so the one case where we might want this is [here](https://github.com/scijava/incubator/pull/95/files#diff-af5dec84a4906407c0159a8ff35d0858d018c6a57302bb6e9c8d81d901107e47R77) - that test will fail because [`OpEnvironment.infos()`](https://github.com/scijava/incubator/pull/95/files#diff-b271cbe48149d64e6e0bfe70e70eadb7d5d3e3f0d83988a06c2e837af4831103R174-R176) will consider the two `OpInfo`s to be the same.

What if we instead changed the `List<String> names()` method of `OpInfo` to return a `Set<String>`, and then we could check set equality in `compareTo` and `equals`?

gselzer replied:
Fixed with 82367790

gselzer commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/OpDescription.java:
```diff
@@ -46,8 +50,8 @@ public final class OpDescription {
      * @param info the {@link OpInfo} of interest
      * @return a descriptor for this {@link OpInfo}
      */
-    public static String basic(final OpInfo info) {
-        return basic(info, null);
+    public static String verbose(final OpInfo info) {
+        return verbose(info, null);
```
I kind of wonder if this  class should just be combined with `SimplifiedOpDescriptionGenerator` somehow...

hinerm replied:
@gselzer it seems odd that the `OpDescription` is a static utility class and like it should be able to be rolled into something else. I don't have a sense for how easily this logic would be accessed in the current places using `OpDescription`. If it's trivial then I would be in favor of refactoring.

gselzer replied:
@hinerm right now it's called in two places:
1) Lots of `OpInfo` implementations use it for their `toString()` - it'd be slick to put it in `OpInfo`, but we can only put `default` methods in there, so that's not ideal
2) The `SimplifiedOpDescriptionGenerator` uses it.

One idea is to combine it with either the `Ops` or the `Infos` utility classes...

gselzer replied:
Migrated all functionality to `Infos.java` in 1a8f51fc

gselzer committed:
First cut of OpEnvironment.help() documentation

gselzer committed:
Update OpEnvironment.help() descriptions

gselzer committed:
Format OpRequest

gselzer committed:
Clean up OpRequest

gselzer committed:
Migrate OpDescription functionality to Infos

gselzer committed:
Document OpDescriptionGenerator

gselzer committed:
Create one discoverer for backend plugins

gselzer committed:
Make a couple methods private

We made a couple of functions protected for reuse in
SimplificationMatchingRoutine, but not all of them needed to be

gselzer committed:
Clean up FocusedOpInfo

gselzer committed:
Remove InfoSimplificationGenerator

gselzer commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/SimplifiedOpDescriptionGenerator.java:
```diff
@@ -0,0 +1,133 @@
+		if (env.infos(name).isEmpty()) {
+			return allNamespaces(env, name);
+		}
+		var infos = SimplificationMatchingRoutine.getSimpleInfos(env, req.getName());
```
Man, I really hate this line. I dislike that we have a separate data structure storing `OpInfo`s that's not in the `DefaultOpEnvironment`, *and* I dislike how the way this is written will ignore all Ops that can't be simplified (which in this case is any `Computer`/`Inplace` whose output can't be copied).

hinerm replied:
>I dislike how the way this is written will ignore all Ops that can't be simplified (which in this case is any Computer/Inplace whose output can't be copied).

@gselzer this sounds particularly concerning.. what're the implications here? Does that mean that these ops won't show up in an `ops.help` call?

gselzer replied:
> > I dislike how the way this is written will ignore all Ops that can't be simplified (which in this case is any Computer/Inplace whose output can't be copied).
> 
> @gselzer this sounds particularly concerning.. what're the implications here? Does that mean that these ops won't show up in an `ops.help` call?

@hinerm yes, if an Op cannot be simplified then it will not show up in an `OpEnvironment.help` call (but it will still show up in an `OpEnvironment.helpVerbose` call)

gselzer replied:
Fixed (albeit HACKy) with 6bcee354

gselzer commented on scijava/scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/simplify/SimplifiedOpDescriptionGenerator.java:
```diff
@@ -0,0 +1,133 @@
+				.filter(n -> n.contains(name)) //
+				.sorted() //
+				.collect(Collectors.toList());
+		return "Namespaces:\n\t> " + String.join("\n\t> ", namespaces);
```
Maybe this should be `OpNames:`

gselzer replied:
Fixed with 8d9373bd

gselzer reviewed (commented):
So close!

gselzer committed:
Enable "simple" descriptions of unSimplified Ops

gselzer committed:
Use "Name" in help string when returning Op names

gselzer committed:
Remove toString() comparison in OpInfo.compareTo
